### PR TITLE
feat(oidc): add allowed scope prefixes

### PR DESCRIPTION
**Pull request by lukaslihotzki-f**
# Which Problems Are Solved

Zitadel currently discards scopes in the authorize request which it cannot process. Therefore, they are omitted in the resulting OIDC session (e.g., in the `oidc_session.added` event).

# How the Problems Are Solved

With this PR, the user can configure the allowed scope prefixes per app, so scopes starting with one of the prefixes are included in the resulting OIDC session. The prefix must contain a colon, so it cannot match plain scopes like `email`, and it cannot start with `urn:zitadel:` or a prefix of it, so it doesn't interfere with zitadel-managed scopes.

No further permission checks are implemented. If necessary, users could add checks with Actions V2.

# Additional Changes

- refactor cnsl-additional-origins so it can also be used also for the allowed scope prefixes

# Additional Context

Related to https://github.com/zitadel/zitadel/issues/8967, but this PR is about accepting client-supplied scopes, and not about adding unrequested scopes automatically. Alternatively, the Device Authorization Grant currently accepts all custom flows. Would that also work for the Authorization Code Grant?

### CODE REVIEW

**Comment by elinashoko** [comment]
Thank you for your contribution @lukaslihotzki-f we'll get to this soon!

File: internal/command/project_application_oidc_test.go [hunk 1]
```diff
@@ -1209,6 +1218,7 @@ func TestCommandSide_ChangeOIDCApplicationSecret(t *testing.T) {
 								"",
 								domain.LoginVersionUnspecified,
 								"",
+								[]string{},
```
Comment (stebenz) [inline internal/command/project_application_oidc_test.go:1363]: Please add test to set the list of allowed scope prefixes and also to set the list to empty again.

File: internal/domain/application_oidc.go [hunk 2]
```diff
@@ -163,6 +164,15 @@ func (a *OIDCApp) OriginsValid() bool {
 	return true
 }
 
+func (a *OIDCApp) AllowedScopePrefixesValid() bool {
+	for _, prefix := range a.AllowedScopePrefixes {
+		if !strings.Contains(prefix, ":") || strings.HasPrefix(prefix, "urn:zitadel:") || strings.HasPrefix("urn:zitadel:", prefix) {
```
Comment (stebenz) [inline internal/domain/application_oidc.go:169]: Not required, but maybe there should be an error related to the validity of the application configuration, to provide information on why the application was not created for example.

File: internal/query/projection/app.go [hunk 3]
```diff
@@ -133,6 +134,7 @@ func (*appProjection) Init() *old_handler.Check {
 			handler.NewColumn(AppOIDCConfigColumnBackChannelLogoutURI, handler.ColumnTypeText, handler.Nullable()),
 			handler.NewColumn(AppOIDCConfigColumnLoginVersion, handler.ColumnTypeEnum, handler.Nullable()),
 			handler.NewColumn(AppOIDCConfigColumnLoginBaseURI, handler.ColumnTypeText, handler.Nullable()),
+			handler.NewColumn(AppOIDCConfigColumnAllowedScopePrefixes, handler.ColumnTypeTextArray, handler.Nullable()),
```
Comment (stebenz) [inline internal/query/projection/app.go]: Column is not defined in the setup step as `NULL`.

File: proto/zitadel/app.proto [hunk 4]
```diff
@@ -179,6 +179,12 @@ message OIDCConfig {
             description: "Specify the preferred login UI, where the user is redirected to for authentication. If unset, the login UI is chosen by the instance default.";
         }
     ];
+    repeated string allowed_scope_prefixes = 23 [
+        (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
+            example: "[\"urn:thirdparty:\"]";
+            description: "prefixes of scopes that are automatically allowed for this application";
```
Comment (stebenz) [inline proto/zitadel/app.proto]:
```suggestion
            description: "Prefixes of scopes that are automatically allowed for this application and handled through the OIDC flow.";
```

File: proto/zitadel/app/v2beta/app_service.proto [hunk 5]
```diff
@@ -595,6 +595,12 @@ message CreateOIDCApplicationRequest {
             description: "Specify the preferred login UI, where the user is redirected to for authentication. If unset, the login UI is chosen by the instance default.";
         }
     ];
+    repeated string allowed_scope_prefixes = 18 [
+        (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
+            example: "[\"urn:thirdparty:\"]";
+            description: "prefixes of scopes that are automatically allowed for this application";
```
Comment (stebenz) [inline proto/zitadel/app/v2beta/app_service.proto]:
```suggestion
            description: "Prefixes of scopes that are automatically allowed for this application and handled through the OIDC flow.";
```

File: proto/zitadel/app/v2beta/app_service.proto [hunk 6]
```diff
@@ -788,6 +794,12 @@ message UpdateOIDCApplicationConfigurationRequest {
...
```
Comment (stebenz) [inline proto/zitadel/app/v2beta/app_service.proto]:
```suggestion
            description: "Prefixes of scopes that are automatically allowed for this application and handled through the OIDC flow.";
```

File: proto/zitadel/app/v2beta/oidc.proto [hunk 7]
```diff
@@ -163,4 +163,10 @@ message OIDCConfig {
             description: "Specify the preferred login UI, where the user is redirected to for authentication. If unset, the login UI is chosen by the instance default.";
         }
     ];
+    repeated string allowed_scope_prefixes = 22 [
+        (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
+            example: "[\"urn:zitadel:iam:org:project:\"]";
+            description: "prefixes of scopes that are automatically allowed for this application";
```
Comment (stebenz) [inline proto/zitadel/app/v2beta/oidc.proto]:
```suggestion
            description: "Prefixes of scopes that are automatically allowed for this application and handled through the OIDC flow.";
```

File: proto/zitadel/management.proto [hunk 8]
```diff
@@ -9954,6 +9954,12 @@ message AddOIDCAppRequest {
             description: "Specify the preferred login UI, where the user is redirected to for authentication. If unset, the login UI is chosen by the instance default.";
         }
     ];
+    repeated string allowed_scope_prefixes = 20 [
+        (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
+            example: "[\"urn:thirdparty:\"]";
+            description: "prefixes of scopes that are automatically allowed for this application";
```
Comment (stebenz) [inline proto/zitadel/management.proto]:
```suggestion
            description: "Prefixes of scopes that are automatically allowed for this application and handled through the OIDC flow.";
```

File: proto/zitadel/management.proto [hunk 9]
```diff
@@ -10145,6 +10151,12 @@ message UpdateOIDCAppConfigRequest {
             description: "Specify the preferred login UI, where the user is redirected to for authentication. If unset, the login UI is chosen by the instance default.";
         }
     ];
+    repeated string allowed_scope_prefixes = 19 [
+        (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
+            example: "[\"urn:thirdparty:\"]";
+            description: "prefixes of scopes that are automatically allowed for this application";
```
Comment (stebenz) [inline proto/zitadel/management.proto]:
```suggestion
            description: "Prefixes of scopes that are automatically allowed for this application and handled through the OIDC flow.";
```

File: internal/api/grpc/app/v2beta/convert/oidc_app.go [hunk 10]
```diff
@@ -38,6 +38,7 @@ func CreateOIDCAppRequestToDomain(name, projectID string, req *app.CreateOIDCApp
 		BackChannelLogoutURI:     gu.Ptr(req.GetBackChannelLogoutUri()),
 		LoginVersion:             loginVersion,
 		LoginBaseURI:             loginBaseURI,
+		AllowedScopePrefixes:     req.GetAllowedScopePrefixes(),
```
Comment (stebenz) [inline internal/api/grpc/app/v2beta/convert/oidc_app.go:41]: There are unit tests for these converts, please add this attribute there as well.

**Review by stebenz — CHANGES_REQUESTED** [review]
PR looks good, but I don't see any integration tests regarding this change.
Will they be added or are there problems setting it up?

**Comment by elinashoko** [comment]
Heya @lukaslihotzki-f can you pls have a look at the requested changes?

**Comment by pkarc** [comment]
Hello everyone,

I've been looking into how to add custom scopes when requesting an OAuth token with a JWT key flow(https://zitadel.com/docs/guides/integrate/service-users/private-key-jwt). I noticed that, as described by @lukaslihotzki-f , the requested scopes were being omitted.

I tried his branch and realized that the changes only apply to apps with an OIDCConfig. In my case, I'm working with an API app, which uses a different configuration struct and a separate table to store this information. So, I created a branch of my own, adapting @lukaslihotzki-f's approach for API apps.

Now I have a question: Do these OIDC flows also apply to API apps? I'm a bit confused about how API apps interact with some parts of the OIDC flows while having their own distinct configurations.

@elinashoko @stebenz, I hope you can shed some light on this and help me figure out if it makes sense to make this work for API apps as well.

Thanks!

**Comment by stebenz** [comment]
Hi @lukaslihotzki-f 
Are you still working on completing this PR as there are some unresolved comments?

Hi @pkarc 
API applications are in general also based on OIDC, but as the configuration differs, which is added for the OIDC applications here, you would need to add it for API applications as well.

**Comment by IAM-marco** [comment]
@lukaslihotzki-f are you still working on this?

**Comment by elinashoko** [comment]
Closing, please reopen if you're still willing to finish this. Thanks!
